### PR TITLE
Fix indent rule to allow for lining up var declarations

### PR DESCRIPTION
**Pull request by ghost**
Our indentation specification is for 3 spaces. We also specify that var declarations need to line up. The base eslint indent rule doesn't allow for this.

This fork of the indent rule will enforce our rules using the following:
`"silvermine/indent": [ "error", 3, { "VariableDeclaratorOffset": { "var": 1, "let": 1, "const": 3 }, "SwitchCase": 1 } ]`

Invalid example:

``` javascript
var a = 1,
   b = 2;
```

Valid example:

``` javascript
var a = 1
    b = 2;
```

### CODE REVIEW

File: lib/rules/indent.js [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+*
+*/
+
+'use strict';
+
+//------------------------------------------------------------------------------
+// Rule Definition
+//------------------------------------------------------------------------------
+var util = require('util'),
+lodash = require('lodash');
```
Comment (jthomerson) [inline lib/rules/indent.js]: irony

File: lib/rules/indent.js [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+* This modification adds the `VariableDeclaratorOffset` configuration variable
+* that allows specifying how many spaces a multi-line variable declaration should
+* be offset from the normal indentation level. For example:
+*
+* var something,
+*     anotherSomething,
+*     somethingElse;
+*
+* The multi-line variable declaration is indented 4 spaces from the normal indentation level,
+* so VariableDeclaratorOffset should be set to '4'.
```
Comment (jthomerson) [inline lib/rules/indent.js]: Is this correct? It's confusing that it says four here and the PR description shows an example using `1`. Probably best to put example configuration right here in the comment.

File: tests/lib/rules/indent.test.js [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+);
+
+validExample = formatCode(
+   'function test() {',
+   '   var a = 1,',
+   '       b = 2;',
+   '',
+   '   return a + b;',
+   '}'
+);
```
Comment (jthomerson) [inline tests/lib/rules/indent.test.js:60]: Please add more invalid and valid examples, including our fluent-chaining examples. This rule needs to check _all_ indentation standards, not just for `var` declarations. So it should make sure that statements in a function are indented one from a function, etc. Think of complex nesting situations and try them here with valid and invalid examples and make these tests more useful and not so surface level. Thanks.

**Comment by jthomerson** [comment]
@jimjenkins5 back to you

File: tests/lib/rules/indent.test.js (hunk 3 above)
Comment (jthomerson) [reply tests/lib/rules/indent.test.js:60]: For example, also include a switch statement.

**Comment by jandmdigital** [comment]
Added more tests and fixed comment.

The VariableDeclaratorOffset should be 1. It's an addition to the indent, not a replacement.

**Comment by jandmdigital** [comment]
@jthomerson back to you

File: tests/lib/rules/indent.test.js [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+   '      B = 2;'
+);
+
+constInvalid = formatCode(
+   'const A = 1,',
+   '   B = 2;',
+   '',
+   'const C = 3,',
+   '    D = 4;'
+)
```
Comment (yokuze) [inline tests/lib/rules/indent.test.js]: Missing a semi-colon here. Please run `grunt standards` before issuing a PR to ensure that the repo meets its own standards.

File: lib/rules/indent.js [hunk 5]
```diff
@@ -0,0 +1,832 @@
+/**
```
Comment (yokuze) [inline lib/rules/indent.js:1]: Running this new indent rule against this repo reveals a lot of indentation errors, both in this new file and in a couple of older files. Please fix these. 

You can run this rule against the repo by using `npm link` and then `npm link eslint-plugin-silvermine` to link this repo to itself (repeating the process for a version of `eslint-config-silvermine` that contains configuration that enables this rule) and then running `grunt standards`.

File: lib/rules/indent.js [hunk 6]
```diff
@@ -0,0 +1,832 @@
...
+                  ) {
+                     nodeIndent = nodeIndent + indentSize;
+                  }
+               }
+            } else if (
+               !parentVarNode && !isFirstArrayElementOnSameLine(parent) &&
+               effectiveParent.type !== 'MemberExpression' &&
+               effectiveParent.type !== 'ExpressionStatement' &&
+               effectiveParent.type !== 'AssignmentExpression' &&
+               effectiveParent.type !== 'Property') {
```
Comment (yokuze) [inline lib/rules/indent.js]: Not only does this block break it's own indentation rule and is difficult to read and understand, it's just plain ugly. Could we perhaps refactor the conditions into a variable or two that will fit onto one line in the if statement?

i.e.

```
isConditionOne = !parentVarNode && !isFirstArrayElementOnSameLine(parent);
isConditionTwo = (['MemberExpression', 'ExpressionStatement', 'AssignmentExpression', 'Property'].indexOf(effectiveParent.type) === -1);
...
} else if (isConditionOne && isConditionTwo) {
```

with better variable names, of course :)

I'm also opening another issue in the `eslint-config-silvermine` repo to disallow multi-line `if` statements, so please eliminate those throughout this file.

**Comment by yokuze** [comment]
Overall, the plugin seems to work well! We just have some formatting issues to take care of. Getting close! @jandmdigital back to you.

File: lib/rules/indent.js (hunk 6 above)
Comment (yokuze) [reply lib/rules/indent.js]: Actually, let's do the same for multi-line `while` and `for` conditionals.

Comment (yokuze) [reply lib/rules/indent.js]: See: https://github.com/silvermine/eslint-config-silvermine/issues/6

**Comment by yokuze** [comment]
@jthomerson this looks good now.

**Pull request merged**